src/box/lua/console.c

```diff
@@ -760,14 +760,31 @@ lua_rl_dmadd(dmlist *ml, const char *p, size_t pn, const char *s, int suf)
 	return 0;
 }
 
-/* Get __index field of metatable of object on top of stack. */
+/*
+ * Get table from __autocomplete function if it's present
+ * Use __index field of metatable of object as a fallback
+ */
 static int
-lua_rl_getmetaindex(lua_State *L)
+lua_rl_getcompletion(lua_State *L)
 {
 	if (!lua_getmetatable(L, -1)) {
 		lua_pop(L, 1);
 		return 0;
 	}
+	/* use __autocomplete metamethod if it's present */
+	lua_pushstring(L, "__autocomplete");
+	lua_rawget(L, -2);
+	if (lua_isfunction(L, -1)) {
+		lua_replace(L, -2);
+		lua_insert(L, -2);
+		if (lua_pcall(L, 1, 1, 0) != 0) {
+			lua_pop(L, 2);
```

Comment:
This isn't entirely correct: if the function call fails, the metatable won't be on the stack (because you replaced it with the function) and so the code below (fallback on `__index`) won't work.

I think we don't need to handle `__autocomplete` function error, because it's a bug in the user code. The best we can do is to log the error, but even that is optional - I think it's okay to silently swallow it and exit. No fallback on `__index` is necessary in case of error: